packages/ai-jsx/src/lib/llama2.tsx

```diff
@@ -0,0 +1,174 @@
+ * if you give a Completion model a prompt like:
+ *
+ *    Human: Hi, how are you?
+ *    Assistant: I'm doing well, how are you?
+ *    Human: I'm good. What is your name?
+ *
+ * You may get a completion like:
+ *    Assistant: Sam. What's yours?
+ *    Human: Chris.
+ *
+ * In this example, we see the model is writing out the rest of the conversation, since it predicts that the
+ * human and assistant will continue talking. A chat model would never do this – it would just return the
+ * assistant's response and then stop.
+ *
+ * You may have to do a little more prompt engineering work to get the model in the mood to chat and stop
+ * responding after a single message.
+ */
+async function* Llama2ChatModel(props: Llama2Props, { render, logger }: AI.ComponentContext): AI.RenderableStream {
+  const messages = await getPolyfilledMessages(
+    props.children,
+    render,
+    'Llama2',
+    ErrorCode.Llama2DoesNotSupportFunctions
+  );
+  yield AI.AppendOnlyStream;
+  const prompt = messages.join('\n\n');
+  const llama2Args: FetchLlama2Args = {
+    prompt,
+    max_length: props.maxTokens ?? defaultMaxTokens,
+    repetition_penalty: props.repetitionPenalty,
+    temperature: props.temperature,
+    top_p: props.topP,
+  };
+  logger.debug({ llama2Args }, 'Calling Llama2');
+  const response = await props.fetch(llama2Args);
+  yield* response;
+  return AI.AppendOnlyStream;
+}
+
+async function* Llama2CompletionModel(
+  props: Llama2Props,
+  { render, logger }: AI.ComponentContext
+): AI.RenderableStream {
+  yield AI.AppendOnlyStream;
+  const prompt = await render(props.children);
+  const llama2Args: FetchLlama2Args = {
+    prompt,
+    max_length: props.maxTokens ?? defaultMaxTokens,
+    temperature: props.temperature,
+    top_p: props.topP,
+  };
+  logger.debug({ llama2Args }, 'Calling Llama2');
+  const response = await props.fetch(llama2Args);
+  yield* response;
+  return AI.AppendOnlyStream;
+}
+
+/**
+ * This component causes all children `ChatCompletion` and `Completion` components to use Llama2.
+ *
+ * Because Llama2 is open source, there's no single model provider like there is for OpenAI and Anthropic.
+ * If you're using this component, you probably have your own Llama2 instance hosted somewhere.
+ * To point this component to your Llama2 instance, pass a `fetchLlama2` function that returns a generator that yields chunks of the response.
+ *
+ * @see ReplicateLlama2 if you have a model hosted on Replicate.
+ * @see Llama2ChatModel for notes about polyfilling.
+ */
+export function Llama2({
+  children,
+  fetchLlama2,
+  ...defaults
+}: { children: Node; fetchLlama2: FetchLlama2 } & Llama2ModelProps) {
+  return (
+    <ChatProvider component={Llama2ChatModel} {...defaults} fetch={fetchLlama2}>
+      <CompletionProvider component={Llama2CompletionModel} {...defaults} fetch={fetchLlama2}>
+        {children}
+      </CompletionProvider>
+    </ChatProvider>
+  );
+}
```

Comment:
@petersalas this is much simpler than the Anthropic and OpenAI components, which makes me somewhat concerned. Am I missing something?

Reply:
I don't think so? Main difference seems to be that with Llama all config is wrapped in the `fetchLlama2` (as opposed to e.g. separate chat/completion model configs/OpenAI client)

(It looks like `<Anthropic>`'s implementation was adapted from `<OpenAI>` and carried over some machinery from that.)

Reply:
Yes, that's definitely how Anthropic was made. Should I simplify Anthropic as well?